### ISSUE TITLE
v8: On second+ runs, repeatedly showing syncing same 2000 blocks

### ISSUE THREAD

**Issue by jamespharvey20** [bitpay/bitcore#1866]
After completing a bitcoin core 0.16.3 initial sync, cloning git d5ce6fa, running `npm install` and `npm start`, bitcore synced up to around block 185581 (I didn't record the exact block number here.)  I needed to reboot the machine, so pressed `CTRL+C`, to what was running from `npm start`, hoping that would lead to a clean shutdown.

After rebooting, I ran `npm start` (from within `bitcore.git/packages/bitcore-node`.)  It started off with `height=185581`.  I checked 1.5 hours later, and saw it was at `186826`.  Knowing that it shouldn't be going that slow, I scrolled up and saw it had kept looping through the same blocks.

Note this second start didn't show the error in #1865.

Note the first start seemed to correctly only show an increasing height number.

In the terminal scrollback over the 1.5 hour period, I see 17,000 lines like this:

```
info: Connected to peer 127.0.0.1 chain=BTC, network=mainnet                                                                                                                    
info: This worker is now the syncing node for BTC mainnet                                                                                                                       
info: Syncing 2000 blocks for BTC mainnet
info: Sync  chain=BTC, network=mainnet, height=185581
info: Sync  chain=BTC, network=mainnet, height=185582
info: Sync  chain=BTC, network=mainnet, height=185583
...
info: Sync  chain=BTC, network=mainnet, height=187578
info: Sync  chain=BTC, network=mainnet, height=187579
info: Sync  chain=BTC, network=mainnet, height=187580
info: Syncing 2000 blocks for BTC mainnet
info: Sync  chain=BTC, network=mainnet, height=185581
info: Sync  chain=BTC, network=mainnet, height=185582
info: Sync  chain=BTC, network=mainnet, height=185583
...
(this repeats around 10 times)
...
info: Sync  chain=BTC, network=mainnet, height=186824
info: Sync  chain=BTC, network=mainnet, height=186825
info: Sync  chain=BTC, network=mainnet, height=186826
^C
```

So, I pulled current git (5 new commits than what I started with) to b42c0adf, ran `npm install` (in `bitcore.git`), and ran `npm start` (in `bitcore.git/packages/bitcore-node`) for my third time, hoping it was a fluke:

```
info: Connected to peer 127.0.0.1 chain=BTC, network=mainnet
info: This worker is now the syncing node for BTC mainnet
info: Syncing 2000 blocks for BTC mainnet
info: Sync  chain=BTC, network=mainnet, height=206827
info: Sync  chain=BTC, network=mainnet, height=206828
info: Sync  chain=BTC, network=mainnet, height=206830
...
info: Sync  chain=BTC, network=mainnet, height=208824
info: Sync  chain=BTC, network=mainnet, height=208825
info: Sync  chain=BTC, network=mainnet, height=208826
info: Syncing 2000 blocks for BTC mainnet
info: Sync  chain=BTC, network=mainnet, height=206827
info: Sync  chain=BTC, network=mainnet, height=206828
info: Sync  chain=BTC, network=mainnet, height=206829
```

So, it happened again on current git.

At its start here, that's about the height I would have expected it to be at, if it hadn't looped around 2000 blocks somewhere around 10 times, which makes me think the second run somehow got in a state where after the `2000 blocks` line, it is reprinting the height number based off the one it started at on that run.  (Although the first example I list here shows one height per line, sometimes it does multiples, so the 17,000 terminal lines I think roughly corresponds to the 21,246 difference in the heights showing at the beginning of the runs.)

I'm hoping it synced those blocks correctly, and is just a display issue.

Again, on my initial run, it was showing the correct numbers.  (I had been charting the time and height number to see how long it's taking, every 30 minutes or so, to make guesses at how long is left, so there's no way I just didn't notice this happening the first run.)  It's only on subsequent runs that it's doing this.

**Comment by micahriggan** [comment]
Thanks for the report, I've started a fresh sync to check this. I'll try to kill it around the same block height.

If it says "Syncing 2000 blocks" and the heights are regressing, that means that it sent it's best 2000 headers to a node, and received the previous 2000 headers back as being needed.

**Comment by micahriggan** [comment]
I've been able to reproduce this.

You can check integrity with this

```
CHAIN=BTC NETWORK=mainnet npm run test:verify
```

**Comment by micahriggan** [comment]
This appears to be an issue with getLocalTip on internal.ts

**Comment by jamespharvey20** [comment]
> Thanks for the report, I've started a fresh sync to check this. I'll try to kill it around the same block height.
> 
> If it says "Syncing 2000 blocks" and the heights are regressing, that means that it sent it's best 2000 headers to a node, and received the previous 2000 headers back as being needed.

Not following if that suggests it's syncing correctly, or if it's not doing anything useful.


> I've been able to reproduce this.
> 
> You can check integrity with this
> 
> ```
> CHAIN=BTC NETWORK=mainnet npm run test:verify
> ```

I hit `CTRL+C` on the running `npm start` which was claiming to loop between `319875` and `321874`.  I ran that command, and got:

```
info: verifying blocks                                                                                                                                                
info: verifying block 0000000000000000156f66579350212dbcd6321ec255d576e21ddc18afc43629: 326884                                                                        
info: verifying block 000000000000000004ddfbc1a6fc9acaeff75b14819f9a97c50202d53171eec5: 326883                                                                        
...
info: verifying block 00000000000000001015f16818648b03a1616bb63cf037aa682fa75aefbd4d0a: 326618                                                                        
info: verifying block 000000000000000019c7a36ec8e970f31edd7010de89e104753cbc07846537ed: 326617                                                                        
^C
```

Since it's just showing `verifying block` without any type of error messages, I'm assuming this means it's syncing correctly with `bitcoind` and is just displaying that it isn't, since those block numbers in the verify weren't ever listed by sync.  Since it was running at about the same speed as syncing, I stopped there and went back to syncing.

I noticed during the `verify`, `bitcoind` and `node` are running at about 50% CPU usage, and `mongod` at about 5%.

Switching back to `npm start`, I see `mongod` running about 400-550% CPU usage, one of the `node` instances at about 85-90%, and `bitcoind` at around 2-3%.  I'm assuming it must be normal for the initial bitcore sync to take a few days after the `bitcoind` sync has completed, but please let me know if there's any settings I can change to let it use more CPU or anything else to speed it up.  It's a 32 logical core (Dual Xeon E5-2670) system with 64GB memory, with both `bitcoind` and `mongod` running on a Samsung 960 EVO 1TB NVMe.  I don't know the parallelization capabilities of the math involved, but I can't help but think perhaps if multiple `node` instances were working, or if the `node` instance doing a lot was multithreaded, that it would go a lot faster.

**Comment by micahriggan** [comment]
It's syncing correctly, it was an issue with getLocalTip. The current height comes from there, it wasn't being updated as the blocks were added.

Mongo should be the main bottleneck, the node process is fetching blocks over p2p, and converting them into mongo updates. If you can increase the mongo save speed of coins, then that'd make the most impact.

As for making it faster, I'm open to suggestions. We're limited to serial processing of blocks by the coin minting / spending process (marking inputs spent, so the inputs must exist). We could create a setting to ignore this, and allow block processing out of order, but I'm not sure how much faster that would make it.

**Comment by jamespharvey20** [comment]
Confirming this fixed things for me.

Regarding speed, it looks to me like you were absolutely right, that Mongo is the main bottleneck.  More specifically, `iostat -xmt 1` shows my Samsung 960 EVO 1TB is most often pegged near 100% usage from Mongo writes.  I tried a RAID0 with a second 960 EVO, but was surprised to see only about a 10% improvement.

**Comment by micahriggan** [comment]
Wow that's surprising to hear about the RAID0. Did it increase the sync speed at least? I would have expected it to continue to max out write speed, but get a bit faster blocks per second wise. There's a benchmark script you can run to see if adding blocks is faster with different configurations.
It's in test/benchmark

**Comment by osagga** [comment]
@jamespharvey20 You can simply do `npm run benchmark` within the `bitcore-node` directory.

**Comment by jamespharvey20** [comment]
> @jamespharvey20 You can simply do `npm run benchmark` within the `bitcore-node` directory.

@osagga Thanks!

> Regarding speed, it looks to me like you were absolutely right, that Mongo is the main bottleneck. More specifically, `iostat -xmt 1` shows my Samsung 960 EVO 1TB is most often pegged near 100% usage from Mongo writes...

I was wrong here.  Turns out, I have been unable to determine if my NVMe is the bottleneck.  `man iostat` says:

"%util... Device saturation occurs when this value is close to 100% for devices serving requests serially.  But for devices serving requests in parallel, such as RAID arrays and modern SSDs, this number does not reflect their performance limits."

> Wow that's surprising to hear about the RAID0. Did it increase the sync speed at least? I would have expected it to continue to max out write speed, but get a bit faster blocks per second wise. There's a benchmark script you can run to see if adding blocks is faster with different configurations.
> It's in test/benchmark

@ micahriggan - I went back and retried different things, and wound up at about the same 10% maximum improvement.  My original try days ago was using a BTRFS RAID0 (natively, not using mdadm.)  I was of course turning off its copy-on-write feature for that volume, as it's often said that helps performance for databases.

Running bitcoind and bitcore on a single NVMe, the benchmark shows for the 80 (1 MB) blocks: 104.051s; 0.769MB/s; and 1.301s/block.  Changing bitcore to run on a mdadm RAID0, with XFS on top, lowered it to: 100.858s; 0.793MB/s; and 1.261s/block.  Changing to running mongodb through numactl lowered it to: 98.935s.  Splitting bitcoind and bitcore off so they're on separate NVME's with no RAID involved, still through numactl, lowered it to: 95.648s; 0.836 MB/s; and 1.196s/block.

I'm around block 537000, so am going to just let it finish up, and not try any more speedups.
